### PR TITLE
Apache Kafka integration as template

### PR DESCRIPTION
**Pull request by craicoverflow**
Adds a Graphback start template that uses Apache Kafka and MongoDB

## Types of changes

What types of changes does your code introduce to Graphback?

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aerogear/graphback/blob/master/CONTRIBUTING.md) document.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added necessary documentation (if appropriate)

## Verification

Refer to the template README for running it. This will also help to verify the README is correct

### CODE REVIEW

File: templates/ts-apollo-kafka-mongo-backend/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  kafka:
+    image: wurstmeister/kafka
+    ports:
+      - 9092:9092
+    links:
+      - zookeeper
+    environment:
+      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
+      KAFKA_ADVERTISED_HOST_NAME: ${HOST_IP}
+      KAFKA_AUTO_CREATE_TOPICS_ENABLE: 'false'
```
Comment (machi1990) [inline templates/ts-apollo-kafka-mongo-backend/docker-compose.yml:18]:
```suggestion
      KAFKA_AUTO_CREATE_TOPICS_ENABLE: 'true'
```
?

Comment (craicoverflow) [reply templates/ts-apollo-kafka-mongo-backend/docker-compose.yml:18]: Thanks, I had that in for testing fix in graphql-kafka-subscriptions

**Comment by wtrocki** [comment]
Priotity: Yesterday :)

File: templates/ts-apollo-mongo-apache-kafka-backend/README.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
+## Graphback MongoDB + Apollo Server with Apache Kafka Template
```
Comment (wtrocki) [inline templates/ts-apollo-mongo-apache-kafka-backend/README.md:1]: Add some note:

> NOTE: this template is available as a preview feature. We looking for feedback and still enhancing it. 

Something like it.

File: templates/ts-apollo-mongo-apache-kafka-backend/README.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+A template that provides you with an easy setup for your application's backend using MongoDB and Apollo server with TypeScript, and Apache Kafka for subscriptions.
```
Comment (wtrocki) [inline templates/ts-apollo-mongo-apache-kafka-backend/README.md:3]:
```suggestion
A template that provides you with an easy setup for your application's backend using MongoDB and Apollo server with TypeScript, and Apache Kafka for GraphQL subscriptions.
```

File: templates/ts-apollo-mongo-apache-kafka-backend/README.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+### Quickstart Guide
+
+The project has been created using Graphback. Run the project using the following steps:
+
+- Start the database
+
+```shell
+HOST_IP=your-host-ip-address && docker-compose up -d
```
Comment (wtrocki) [inline templates/ts-apollo-mongo-apache-kafka-backend/README.md:12]: This is some linux stuff. On mac that host will be always locahost.

File: templates/ts-apollo-mongo-apache-kafka-backend/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,31 @@
+version: '3'
+services:
+
+  zookeeper:
+    image: wurstmeister/zookeeper
+    ports:
+      - 2181:2181
+
+  kafka:
+    image: wurstmeister/kafka
```
Comment (wtrocki) [inline templates/ts-apollo-mongo-apache-kafka-backend/docker-compose.yml:10]: Where this is comming from?

`image: apache-kafka/broker:2.5.0` 

We should use kafka from strimzi or if you do not like strimzi then just use official sources. 
This is some guy there. Just swap it from docker-compose from kafka website.

File: templates/ts-apollo-mongo-apache-kafka-backend/src/util.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+export function getKafkaTopicName(modelName: string): string {
+  const applicationName = process.env.APPLICATION_NAME || 'graphback-kafka-template-app';
+
+  return `${applicationName}.${modelName}`;
```
Comment (wtrocki) [inline templates/ts-apollo-mongo-apache-kafka-backend/src/util.ts:4]: `_` or `-`  is preferred due to clash with ACL/auth schemes (and web formats)

Comment (wtrocki) [reply templates/ts-apollo-mongo-apache-kafka-backend/src/util.ts:4]: This is actually good delimiter etc. just having context `-` will be better

**Review by wtrocki — APPROVED** [review]

File: templates/ts-apollo-mongo-apache-kafka-backend/docker-compose.yml (hunk 5 above)
Comment (craicoverflow) [reply templates/ts-apollo-mongo-apache-kafka-backend/docker-compose.yml:10]: I chose this because it is by far the most used Kafka image on Docker (> 10M pulls)

https://hub.docker.com/r/wurstmeister/kafka

I could not find a better alternative.
```
$ docker pull apache-kafka/broker
Using default tag: latest
Error response from daemon: pull access denied for apache-kafka/broker, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

https://hub.docker.com/r/strimzi/kafka

strimzi/kafka has no documentation on how to set it up - I think it is only intended for direct use with the Strimzi Operator.

We _could_ figure out how to use it by looking at the operator documentation, but this just makes it almost impossible for a regular user to make changes without also having to do the same.

File: templates/ts-apollo-mongo-apache-kafka-backend/src/util.ts (hunk 6 above)
Comment (craicoverflow) [reply templates/ts-apollo-mongo-apache-kafka-backend/src/util.ts:4]: Sure - I was doing some reading on naming conventions and came to this conclusion, but easy change :)

Comment (craicoverflow) [reply templates/ts-apollo-mongo-apache-kafka-backend/src/util.ts:4]: Blog post: https://devshawn.com/blog/apache-kafka-topic-naming-conventions/

TL;DR something like `<data-center>.<domain>.<classification>.<description>.<version>`

Comment (wtrocki) [reply templates/ts-apollo-mongo-apache-kafka-backend/src/util.ts:4]: ok.. with that as well. I will see how this will work with auth later. so let;'s merge this this way.
